core/src/networked_client.cpp

```diff
@@ -0,0 +1,219 @@
+#include <knoting/engine.h>
+#include <knoting/networked_client.h>
+
+namespace knot {
+using namespace yojimbo;
+NetworkedClient::~NetworkedClient() {}
+
+NetworkedClient::NetworkedClient(Engine& engine) : m_engine(engine), m_client(nullptr) {}
+
+void NetworkedClient::on_awake() {
+    m_client = std::make_shared<Client>(GetDefaultAllocator(), Address("0.0.0.0", CLIENT_PORT), m_config, gameAdapter,
+                                        get_time());
+    m_inManager = m_engine.get_window_module().lock()->get_input_manager();
+    log::debug("CREATED CLIENT with {} channels", m_config.numChannels);
+}
+void NetworkedClient::on_update(double m_delta_time) {
+    if (m_client->IsDisconnected()) {
+        log::debug("Client is not Connected to Server!");
+        while (!attempt_connection())
+            ;
+        return;
+    }
+    m_client->AdvanceTime(get_time());
+    m_client->ReceivePackets();
+
+    m_tickTime += m_delta_time;
+    if (m_client->IsConnected()) {
+        if (!connected) {
+            connected = true;
+            log::debug("Client {} connected to server", m_clientId);
+        }
+        handle_recieved_packets();
+        test_player_input();
+        send_message();
+    }
+    if (m_tickTime >= TICK) {
+        m_tickTime -= TICK;
+    }
+
+    m_client->SendPackets();
+}
+
+void NetworkedClient::on_fixed_update() {}
+void NetworkedClient::on_late_update() {
+    Subsystem::on_late_update();
+}
+void NetworkedClient::on_destroy() {
+    m_client->Disconnect();
+    m_client = nullptr;
+}
+double NetworkedClient::get_time() {
+    auto engOpt = Engine::get_active_engine();
+    if (engOpt) {
+        return engOpt.value().get().get_current_time();
+    }
+}
+bool NetworkedClient::attempt_connection() {
+    if (m_client->IsConnecting()) {
+        return true;
+    }
+    yojimbo::random_bytes((uint8_t*)&m_clientId, 8);
+    m_serverAddress = Address(SERVER_ADDRESS, SERVER_PORT);
+
+    uint8_t privateNetworkingKey[yojimbo::KeyBytes] = {0};
+    m_client->InsecureConnect(privateNetworkingKey, m_clientId, m_serverAddress);
+    log::debug("ATTEMPTED CLIENT CONNECTION");
+    return m_client->IsConnecting();
+}
+bool NetworkedClient::handle_recieved_packets() {
+    if (!m_client->IsConnected()) {
+        return false;
+    }
+    if (!(m_client)) {
+        return false;
+    }
+    Message* mess = nullptr;
+    while ((mess = m_client->ReceiveMessage(1))) {
```

Comment:
This can be more readable with a do while

Reply:
can't do this due to the way messages are released at the end of the while